src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1395,6 +1405,12 @@ MulticopterPositionControl::task_main()
 				if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
 						&& _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LAND) {
 					_vel_sp(2) = _params.land_speed;
+
+					/* final approach, decrease descent speed */
+					if(_params.land_final_approach_alt > 0.0f && _local_pos.z_valid
```

Comment:
Also, in addition to if the terrain estimate is valid we also need to factor in the terrain estimation velocity and check that it's going down. This will prevent listening to stuck range finder readings that we cannot detect somehow else at the moment (a real problem with the Pulsedlight devices).
But for this we need to add the terrain estimator down velocity to the global position message, coming from here (published from the position estimators): https://github.com/PX4/Firmware/blob/master/src/lib/terrain_estimation/terrain_estimator.h#L65